File: ldk/node/src/hostClients/browserClient.ts
import { Empty } from 'google-protobuf/google/protobuf/empty_pb';
import { BrowserClient as BrowserGRPCClient } from '../grpc/browser_grpc_pb';
import Messages from '../grpc/browser_pb';
import BaseClient, { GRPCClientConstructor } from './baseClient';
import { BrowserSelectedTextResponse, BrowserService } from './browserService';
import { StoppableStream, StreamListener } from './stoppableStream';
import { StreamTransformer, TransformingStream } from './transformingStream';

/**
 * @internal
 *
 * @param message
 */
const transformSelectedTextResponse: StreamTransformer<
  | Messages.BrowserSelectedTextResponse
  | Messages.BrowserSelectedTextStreamResponse,
  BrowserSelectedTextResponse
> = (message) => {
  return {
    url: message.getUrl(),
    text: message.getText(),
    tabTitle: message.getTabtitle(),
  };
};

/**
 * @internal
 */
export class BrowserClient
  extends BaseClient<BrowserGRPCClient>
  implements BrowserService {
  queryActiveURL(): Promise<string> {
    return this.buildQuery<Empty, Messages.BrowserActiveURLResponse, string>(
      (message, callback) => this.client.browserActiveURL(message, callback),
      () => new Empty(),
      (response) => response.getUrl(),
    );
  }

  querySelectedText(): Promise<BrowserSelectedTextResponse> {
    return this.buildQuery<
      Empty,
      Messages.BrowserSelectedTextResponse,
      BrowserSelectedTextResponse
    >(
      (message, callback) => this.client.browserSelectedText(message, callback),
      () => new Empty(),
      transformSelectedTextResponse,
    );
  }

  streamActiveURL(listener: StreamListener<string>): StoppableStream<string> {
    return new TransformingStream<
      Messages.BrowserActiveURLStreamResponse,
      string
    >(
      this.client.browserActiveURLStream(new Empty()),
      (message) => message.getUrl(),
      listener,
    );
  }

  streamSelectedText(
    listener: StreamListener<BrowserSelectedTextResponse>,
  ): StoppableStream<BrowserSelectedTextResponse> {
    return new TransformingStream<
      Messages.BrowserSelectedTextStreamResponse,
      BrowserSelectedTextResponse
    >(
      this.client.browserSelectedTextStream(new Empty()),
      transformSelectedTextResponse,
      listener,
    );
  }

  protected generateClient(): GRPCClientConstructor<BrowserGRPCClient> {
    return BrowserGRPCClient;
  }
}

File: ldk/csharp/OliveHelpsLDK/OliveHelpsLDK/Browser/BrowserClient.cs
using System.Threading;
using System.Threading.Tasks;
using Grpc.Core;
using OliveHelpsLDK.Logging;
using Proto;

namespace OliveHelpsLDK.Browser
{
    internal class BrowserClient : BaseClient<Proto.Browser.BrowserClient>, IBrowserService
    {
        internal BrowserClient(Proto.Browser.BrowserClient client, Session session, ILogger logger) : base(
            client, session, logger, "browser")
        {
        }

        internal BrowserClient(CallInvoker channelBase, Session session, ILogger logger) : this(
            new Proto.Browser.BrowserClient(channelBase), session, logger)
        {
        }

        public Task<string> QueryActiveURL(CancellationToken cancellationToken = default)
        {
            var request = new BrowserActiveURLRequest
            {
                Session = CreateSession()
            };
            var loggedParser = LoggedParser<Task<BrowserActiveURLResponse>, string>(task => task.Result.Url);
            return Client.BrowserActiveURLAsync(request, CreateOptions(cancellationToken)).ResponseAsync
                .ContinueWith(loggedParser, cancellationToken, TaskContinuationOptions.OnlyOnRanToCompletion,
                    TaskScheduler.Current);
        }

        public IStreamingCall<string> StreamActiveURL(CancellationToken cancellationToken = default)
        {
            var request = new BrowserActiveURLStreamRequest
            {
                Session = CreateSession()
            };
            var call = Client.BrowserActiveURLStream(request, CreateOptions(cancellationToken));
            var loggedParser = LoggedParser<BrowserActiveURLStreamResponse, string>(resp => resp.Url);
            return new StreamingCall<BrowserActiveURLStreamResponse, string>(call, loggedParser);
        }

        public Task<SelectedText> QuerySelectedText(CancellationToken cancellationToken = default)
        {
            var request = new BrowserSelectedTextRequest
            {
                Session = CreateSession()
            };
            var loggedParser =
                LoggedParser<Task<BrowserSelectedTextResponse>, SelectedText>(task => FromProto(task.Result));
            return Client.BrowserSelectedTextAsync(request, CreateOptions(cancellationToken)).ResponseAsync
                .ContinueWith(loggedParser, cancellationToken, TaskContinuationOptions.OnlyOnRanToCompletion,
                    TaskScheduler.Current);
        }

        public IStreamingCall<SelectedText> StreamSelectedText(CancellationToken cancellationToken = default)
        {
            var request = new BrowserSelectedTextStreamRequest
            {
                Session = CreateSession()
            };
            var call = Client.BrowserSelectedTextStream(request, CreateOptions(cancellationToken));
            var loggedParser =
                LoggedParser<BrowserSelectedTextStreamResponse, SelectedText>(response => FromProto(response));
            return new StreamingCall<BrowserSelectedTextStreamResponse, SelectedText>(call,
                loggedParser);
        }

        internal static SelectedText FromProto(BrowserSelectedTextResponse response)
        {
            return new SelectedText
            {
                TabTitle = response.TabTitle,
                Text = response.Text,
                URL = response.Url,
            };
        }

        internal static SelectedText FromProto(BrowserSelectedTextStreamResponse response)
        {
            return new SelectedText
            {
                TabTitle = response.TabTitle,
                Text = response.Text,
                URL = response.Url,
            };
        }
    }
}